vfkit: Add Rosetta support for deploying amd64 images on Apple silicon

nirs opened the pull request:
Add --rosetta flag enabling Rosetta[1] for running apps built for Intel processor on Mac with Apple silicon. An example use case is deploying open-cluster-management that do not provide arm64 builds yet.

If Rosetta is not installed on the host, it will be installed on the first time starting a cluster. When running in non-interactive mode automatic install is disabled and if Rosetta is not installed start will fail.

The --rosetta flag is ignored with a warning if enabled on Mac with Intel processor.

## Testing amd64 executable on Apple silicon

```console
% cat rosetta/test.go 
package main

import (
        "bytes"
        "fmt"
        "os/exec"
        "runtime"
)

func main() {
        uname, _ := exec.Command("uname", "-a").Output()
        fmt.Printf("✅ Running %s binary on %s\n", runtime.GOARCH, bytes.TrimSpace(uname))
}

% GOARCH=amd64 GOOS=linux go build -o rosetta/test rosetta/test.go

% file rosetta/test
rosetta/test: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=9743ccad9a20c7d535a9794cc7b22ed1147c89d8, with debug_info, not stripped

% out/minikube start --driver vfkit --rosetta --mount-string $PWD/rosetta:/mnt/rosetta --no-kubernetes
😄  minikube v1.37.0 on Darwin 26.1 (arm64)
✨  Using the vfkit driver based on user configuration
👍  Starting minikube without Kubernetes in cluster minikube
🔥  Creating vfkit VM (CPUs=2, Memory=6144MB, Disk=20000MB) ...
🐳  Preparing Docker 28.5.2 ...
🏄  Done! minikube is ready without Kubernetes!

% minikube ssh /mnt/rosetta/test                                  
✅ Running amd64 binary on Linux minikube 6.6.95 #1 SMP PREEMPT Thu Dec 11 21:09:52 UTC 2025 aarch64 GNU/Linux
```

## Testing with container images

I tested the PR with ramen regional DR environment, using containerd runtime. We deploy several adm64 only container images like open-cluster-management and ramen.
https://github.com/RamenDR/ramen/pull/2356

---

[1] https://developer.apple.com/documentation/virtualization/running-intel-binaries-in-linux-vms-with-rosetta

Fixes #20559

cfergeau commented on cmd/minikube/cmd/start_flags.go:
```diff
@@ -548,6 +552,24 @@ func validateVfkitNetwork(n string, options *run.CommandOptions) string {
+	enabled := viper.GetBool(rosetta)
+	if enabled {
+		if !driver.IsVFKit(driverName) {
+			out.Styled(style.Warning, "Rosetta only supported on for the vfkit driver")
```
Keep only `on` or `for`, but not both (`on for`) in the warning message.

nirs replied:
Thanks, fixing

cfergeau commented on pkg/drivers/vfkit/vfkit.go:
```diff
@@ -350,6 +365,57 @@ func (d *Driver) startVfkit(socketPath string) error {
+	fmt.Fprintf(&b, "set -e\n")
+	fmt.Fprintf(&b, "sudo mount binfmt_misc -t binfmt_misc /proc/sys/fs/binfmt_misc\n")
+	fmt.Fprintf(&b, "sudo mkdir -p %s\n", rosettaMountPoint)
+	fmt.Fprintf(&b, "sudo mount -t virtiofs %s %s\n", rosettaMountTag, rosettaMountPoint)
```
This could use `virtiofs.SetupMounts`

nirs commented:
@cfergeau I addressed the comments:
- Typos in warnings fixed
- Warnings change to match better similar warnings around this code
- I kept setupRosetta() as is for simplicity and ease of maintainability

obnoxxx commented:
@nirs, with your hint, I got minikube start to complete now, and running the test binary in the vm succeeds:

```console
% ./out/minikube ssh
$ /mnt/rosetta/test
✅ Running amd64 binary on Linux minikube 6.6.95 #1 SMP PREEMPT Tue Dec 16 02:19:13 UTC 2025 aarch64 GNU/Linux
$ 
```

so this seems all fine.

medyagh commented on cmd/minikube/cmd/start_flags.go:
```diff
@@ -285,6 +286,9 @@ func initDriverFlags() {
 	startCmd.Flags().String(qemuFirmwarePath, "", "Path to the qemu firmware file. Defaults: For Linux, the default firmware location. For macOS, the brew installation location. For Windows, C:\\Program Files\\qemu\\share")
+
+	// vfkit
+	startCmd.Flags().Bool(rosetta, false, "Enable Rosetta to support apps built for Intel processor on a Mac with Apple silicon (vfkit driver only)")
```
does it also support running images built for x86 on all container runtimes ?
or is it only for binaries ? if also images, should mention in the help text.

it would be nice to check or verify (
suggestion, add an integration test that pulls an x86 image and runs it on silicon)
and test be skipped on anything but macos arm64
(could be a subset of functional test)

if dont wanna add ftest, can also just add manual verification and add to the PR description.

medyagh commented on pkg/drivers/vfkit/vfkit.go:
```diff
@@ -350,6 +365,57 @@ func (d *Driver) startVfkit(socketPath string) error {
 	return process.WritePidfile(d.pidfilePath(), cmd.Process.Pid)
 }
 
+func (d *Driver) rosettaOptions() []string {
```
could add a short general comment for easy code skimming
